keps/sig-cloud-provider/20200928-extend-konnectivity-for-both-directions.md

```diff
@@ -0,0 +1,194 @@
+
+In order to enable traffic from Kubelets and Pods running on Master Network, the Konnectivity Agents have to expose an endpoint that will be listening on a specific port for each of the destinations on the Master Network. As opposed to the traffic flowing from the Master Network to the Node Network, the Konnectivity Agent should act transparently: From a Kubelets or Pods standpoint, the Konnectivity Agent should be the final destination instead of acting as a proxy. 
+
+The reason why we do not intend to use the same strategy used in the other direction is that we do not have control over the clients using the Kubernetes default service to communicate with the KAS.
+
+### Traffic Flow
+
+```
+client =TCP=> (:6443) agent GRPC=> server =TCP=> KAS(:6443)
+                         |            ^
+                         |   Tunnel   |
+                         +------------+
+```
+
+The agent listens for TCP connections at a specific port for each configured destination. When a connection request is received by the Konnectivity Agent the following happens:
+1. A GRPC DIAL_REQ message is sent to the Konnectivity server containing the destination address associated with the current port.
+2. Upon reception of the DIAL_REQ the Konnectivity Server opens a TCP connection with the destination host/port and replies to the Konnectivity Agent with a GRPC DIAL_RES message.
+3. At this point the tunnel is established and data is piped through it, carried over GRPC DATA packets.
+
+### Agent additional flags
+
+* `--target=local_port:dst_host_ip:dst_port`: We can have multiple of those in order to support multiple destinations on the Master Network.
+Dst_host_ip: end target IP (apiserver or something else). In case of IPv6
+the address should be wrapped in squared brackets, consistently with [RFC3986](https://www.ietf.org/rfc/rfc3986.txt) notation.
+e.g. --target=6443:apiserver.svc.cluster.local:6443
+e.g. --target=6443:[2001:db8:1f70::999:de8:7648:6e8]:6443
+
+* `--bind-address=ip`: Local IP address where the Konnectivity Agent will listen for incoming requests. It will be bound to a dummy IP interface with IP x.y.z.v defined by the user. Must be used with the previous one to enable incoming requests. If not, and for backward compatibility, only the traffic initiated from the Control Plane will be allowed.
+
+### Handling the Traffic from the Pods to the Agent
+
+As mentioned above, pods make use of  the Kubernetes default service to reach the KAS. To keep things transparent from a Pod perspective, they will hit the Konnectivity Agent using the Kubernetes default service. The endpoint will be the Konnectivity Agent instead of the KAS. 
+The configuration part of the Kubernetes default service will be done using the Apiserver flag `--advertise-address ip` on the Control Plane side.
+`--advertise-address ip` should match the `--bind-address ip` of the Konnectivity Agent described above. 
```

Comment:
Would the serving port used in updating the default `kubernetes` Service be updated as well? And if so that would imply kube-apiserver and Konnectivity Agent listen on the same ports?

Reply:
Yes, the Agent and the KAS should listen on the same port. I think that we should make this more clear. Thx for the hint.